Source/FakeItEasy.Tests/Core/ArgumentValueFormatterTests.cs

```diff
@@ -142,6 +143,40 @@ public void Should_prefer_exact_type_formatter_to_interface_formatter()
             Assert.That(result, Is.EqualTo("a string"));
         }
 
+        [Test]
+        public void Should_be_thread_safe_without_corrupting_dictionary()
+        {
+            ArgumentValueFormatter localFormatter = new ArgumentValueFormatter(new IArgumentValueFormatter[0]);
+            var a = new object[] {
+                new { a = 0 },
+                new { b = 0 },
+                new { c = 0 },
+                new { d = 0 },
+                new { e = 0 },
+                new { f = 0 },
+                new { g = 0 },
+                new { h = 0 },
+                new { i = 0 },
+                new { j = 0 },
+                new { k = 0 },
+                new { l = 0 },
+                new { m = 0 },
+                new { n = 0 },
+                new { o = 0 },
+                new { p = 0 },
+                new { q = 0 },
+                new { r = 0 },
+                new { s = 0 },
+                new { t = 0 },
+            };
+
+            Parallel.For(0, 20, _i => 
+                {
+                    localFormatter.GetArgumentValueAsString(a[_i]);
+                });
+            Assert.That(localFormatter.cachedFormatters.Count, Is.EqualTo(20));
+        }
+

```

Comment:
As I mentioned in the conversation, I'm not sold on this approach. In my admittedly unscientific testing, using `Dictionary` for the `cachedFormatters`, I see a failure quite infrequently. Certainly no more than once in every 4 runs. I'll not comment more, waiting for your response about why you think a non-deterministic test is superior.